File: src/resolver_sim/allocation/realized_statement.clj
(ns resolver-sim.allocation.realized-statement
  "Canonical cross-runtime statement for a realized allocation:
   `realized-allocation-statement.v1`.

   This is the semantic object that agent-c produces and commits, that an
   independent Rust kernel reproduces, and that the SP1 guest eventually
   proves. It deliberately does NOT expose agent-c's internal evidence
   ontology: the statement binds a fixed set of roots so a verifier in any
   runtime agrees on the same object without sharing evidence structures.

   The statement binds six roots:

     :allocation-context-root    — committed allocation context (round inputs)
     :request-set-root           — committed requested set per claim
     :allocation-policy-root     — committed allocation/fill policy
     :realized-results-root      — committed realized fills (filled/deferred/haircut)
     :fail-action-policy-root    — committed declared fail-action policy
     :round-lifecycle-root       — committed round-lifecycle projection

   and commits the resulting map under REALIZED_ALLOCATION_STATEMENT_V1.

   All-active no-churn property (mirrors conclusion-collective-hash): when the
   allocation is all-active — no rejection and no deferred/haircut fail action —
   the realized-results root is the unfiltered commitment over the filled set;
   the rejection/fail-action filter is a no-op, so the statement root is
   identical whether or not fail-action filtering was applied.

   Scenario-evidence binding: agent-c binds the statement root into its richer
   scenario evidence via :scenario-evidence-binding, so SP1 proves the canonical
   statement while the simulator's richer evidence model stays simulator-side."
  (:require [resolver-sim.allocation.context :as context]
            [resolver-sim.hash.canonical :as hc]))

(def schema-version "realized-allocation-statement.v1")

(defn- sort-by-claim-id
  "Canonical order for claim-keyed maps: sort by claim id key."
  [m]
  (into (sorted-map) m))

(defn request-set-root
  "Commit the requested set (claim -> amount) under REALIZED_REQUEST_SET_V1."
  [requested]
  (hc/domain-hash :realized-request-set (sort-by-claim-id requested)))

(defn allocation-policy-root
  "Commit the allocation/fill policy under ALLOCATION_POLICY_V1.
   The policy is the effective fill policy (mode, rounding policy, fill
   order, residual treatment); the declared fail-action policy is committed
   separately so a policy change cannot silently change the verifier's meaning
   without changing evidence identity."
  [policy]
  (hc/domain-hash :allocation-policy (dissoc policy :fail-action-policy)))

(defn fail-action-policy-root
  "Commit the declared fail-action policy under FAIL_ACTION_POLICY_V1.
   When no fail-action policy is declared, commits the canonical conservative
   default so the root is always defined and comparable."
  [policy]
  (let [effective (or (:fail-action-policy policy)
                      {:mode :pro-rata-treatment
                       :deferred-policy :same-ratio
                       :haircut-policy :same-ratio})]
    (hc/hash-with-intent {:hash/intent :fail-action-policy} effective)))

(defn disposition-of
  "Classify one participant's realized disposition from its realized amounts.

   Order of precedence: deferred, then haircut, then full/partial/zero fill.
   A participant present in :requested always gets an explicit disposition, so
   'inactive/zero-filled' is distinguishable from 'producer omitted'.

   Public so the Rust mirror can implement the identical classifier and golden
   vectors can pin it."
  [{:keys [requested filled deferred haircut]}]
  (cond
    (and (pos? (long deferred)) (pos? (long haircut))) :deferred-and-haircut
    (pos? (long deferred)) :deferred
    (pos? (long haircut)) :haircut
    (= (long requested) (long filled)) :full-fill
    (zero? (long filled)) :zero-filled
    :else :partial-fill))

(defn realized-results-root
  "Commit the realized results under REALIZED_RESULTS_V1 as an explicit
   per-participant disposition vector.

   The committed projection is a sorted vector of rows over the union of
   requested keys (every participant that had a request), each row carrying:

     {:claim/id k
      :requested r  :filled f  :deferred d  :haircut h  :unrealized u
      :disposition :full-fill | :partial-fill | :deferred | :haircut
                   | :zero-filled | :deferred-and-haircut}

   This is the participant/request → realized-disposition model: no participant
   is silently dropped. 'absent because inactive' (row present, :zero-filled)
   is distinguishable from 'absent because the producer omitted it' (row
   missing entirely), which changes the root.

   All-active no-churn: an all-active allocation is exactly the projection
   where every row is :full-fill with no deferred/haircut — the fail-action
   filter is a no-op, so the all-active root is byte-identical whether or not
   fail-action filtering was applied."
  [{:keys [requested filled deferred haircut unrealized]}]
  (let [claims (->> (concat (keys requested) (keys filled)
                            (keys deferred) (keys haircut) (keys unrealized))
                    distinct
                    sort)]
    (hc/domain-hash :realized-results
                    (mapv (fn [k]
                            (let [row {:claim/id k
                                       :requested (long (get requested k 0))
                                       :filled (long (get filled k 0))
                                       :deferred (long (get deferred k 0))
                                       :haircut (long (get haircut k 0))
                                       :unrealized (long (get unrealized k 0))}]
                              (assoc row :disposition (disposition-of row))))
                          claims))))

(defn round-lifecycle-root
  "Commit the round-lifecycle projection under ROUND_LIFECYCLE_V1."
  [round-lifecycle]
  (hc/domain-hash :round-lifecycle round-lifecycle))

(defn allocation-context-root
  "Commit the allocation context (round inputs) under ALLOCATION_CONTEXT_V1."
  [ctx]
  (context/context-hash ctx))

(defn all-active?
  "True when the realized allocation is all-active: nothing deferred, nothing
   haircut, and every requested amount is fully filled."
  [{:keys [requested filled deferred haircut]}]
  (and (empty? deferred)
       (empty? haircut)
       (= (sort-by-claim-id requested)
          (sort-by-claim-id filled))))

(defn statement-preimage
  "The canonical statement value tree committed by the statement root."
  [{:keys [allocation-context-root request-set-root allocation-policy-root
           realized-results-root fail-action-policy-root round-lifecycle-root]}]
  {:schema-version schema-version
   :allocation-context-root allocation-context-root
   :request-set-root request-set-root
   :allocation-policy-root allocation-policy-root
   :realized-results-root realized-results-root
   :fail-action-policy-root fail-action-policy-root
   :round-lifecycle-root round-lifecycle-root})

(defn statement-root
  "Commit the realized-allocation-statement.v1 under
   REALIZED_ALLOCATION_STATEMENT_V1."
  [statement]
  (hc/domain-hash :realized-allocation-statement (statement-preimage statement)))

(defn build-statement
  "Build a realized-allocation-statement.v1 from a decision and its context.

   Inputs:
     :ctx            — allocation context map (context/build-context output)
     :decision       — partial-fill decision (requested/filled/deferred/haircut
                       + :policy with optional :fail-action-policy)
     :round-lifecycle— round-state/round-lifecycle projection

   Returns the full statement map including :statement/root,
   :statement/all-active? and :statement/verification-equalities."
  [{:keys [ctx decision round-lifecycle]}]
  (let [statement {:schema-version schema-version
                   :allocation-context-root (allocation-context-root ctx)
                   :request-set-root (request-set-root (:requested decision))
                   :allocation-policy-root (allocation-policy-root (:policy decision))
                   :realized-results-root (realized-results-root decision)
                   :fail-action-policy-root (fail-action-policy-root (:policy decision))
                   :round-lifecycle-root (round-lifecycle-root round-lifecycle)}]
    (assoc statement
           :statement/root (statement-root statement)
           :statement/all-active? (all-active? decision)
           :statement/verification-equalities
           {:all-active-all-full-fill
            (let [active? (all-active? decision)
                  disposition (fn [k]
                                (disposition-of
                                 {:requested (long (get-in decision [:requested k] 0))
                                  :filled (long (get-in decision [:filled k] 0))
                                  :deferred (long (get-in decision [:deferred k] 0))
                                  :haircut (long (get-in decision [:haircut k] 0))
                                  :unrealized (long (get-in decision [:unrealized k] 0))}))]
              (and active?
                   (every? #(= :full-fill (disposition %))
                           (keys (:requested decision)))))})))

(defn scenario-evidence-root
  "Bind the realized-allocation-statement root into agent-c's scenario
   evidence via SCENARIO_EVIDENCE_BINDING_V1.

   SP1 proves the canonical statement; the simulator's richer evidence model
   stays simulator-side but is cryptographically linked to the statement, so
   a verifier can confirm the scenario evidence corresponds to the proven
   allocation statement."
  [statement scenario-evidence]
  (hc/domain-hash :scenario-evidence-binding
                  {:realized-allocation-statement-root (:statement/root statement)
                   :scenario-evidence scenario-evidence}))

File: coprocessor/core/src/realized_statement.rs
//! Realized-allocation-statement projection (independent Rust implementation).
//!
//! `realized-allocation-statement.v1`: commits six roots derived from raw
//! inputs, then commits the statement root under
//! REALIZED_ALLOCATION_STATEMENT_V1.
//!
//! This is the projection + canonical-encode layer. It is deliberately
//! decoupled from `realized_fill.rs` (the semantics layer) so serialization
//! and semantic errors are testable independently.
//!
//! Per acceptance condition A, Rust reconstructs all six roots from raw inputs;
//! it does NOT accept pre-computed roots as inputs. Each root function takes
//! the canonical raw inputs and recomputes the root itself.
//!
//! Claim keys are STRINGS (claim-id), matching the allocation context.

use crate::canonical::{domain_hash, tags, CanonValue};
use crate::kernel::{context_preimage, Context};
use crate::lifecycle::RoundLifecycle;
use crate::realized_fill::{canon_request_set, RealizedAllocation};

pub const SCHEMA_VERSION: &str = "realized-allocation-statement.v1";

/// The allocation-context root: domain hash of the context preimage, derived
/// from the raw `Context` (condition A: Rust reconstructs, does not bind a
/// precomputed root). The context itself is the round-level input committed by
/// the kernel and reproduced byte-for-byte by `kernel::context_preimage`.
pub fn allocation_context_root(ctx: &Context) -> String {
    domain_hash(tags::ALLOCATION_CONTEXT, &context_preimage(ctx))
}

/// Request-set root: domain hash of the claim-keyed requested set under
/// REALIZED_REQUEST_SET_V1.
pub fn request_set_root(requested: &[(String, i64)]) -> String {
    domain_hash("REALIZED_REQUEST_SET_V1", &canon_request_set(requested))
}

/// Allocation-policy root: domain hash of the effective fill policy under
/// ALLOCATION_POLICY_V1. The declared fail-action policy is excluded here and
/// committed separately, so a policy change cannot silently change the
/// verifier's meaning without changing evidence identity.
pub fn allocation_policy_root(policy: &[(String, CanonValue)]) -> String {
    let policy_canon = keyword_map(policy);
    domain_hash("ALLOCATION_POLICY_V1", &policy_canon)
}

/// Fail-action-policy root: domain hash of the declared fail-action policy
/// under FAIL_ACTION_POLICY_V1. When no policy is declared, the canonical
/// conservative default is committed so the root is always defined.
pub fn fail_action_policy_root(fail_action_policy: Option<&[(String, CanonValue)]>) -> String {
    let canonical = match fail_action_policy {
        Some(pairs) => keyword_map(pairs),
        None => CanonValue::map(vec![
            (
                CanonValue::keyword("mode"),
                CanonValue::keyword("pro-rata-treatment"),
            ),
            (
                CanonValue::keyword("deferred-policy"),
                CanonValue::keyword("same-ratio"),
            ),
            (
                CanonValue::keyword("haircut-policy"),
                CanonValue::keyword("same-ratio"),
            ),
        ]),
    };
    domain_hash("FAIL_ACTION_POLICY_V1", &canonical)
}

/// Convert (string-key, value) policy pairs to a canonical map with keyword
/// keys.
fn keyword_map(pairs: &[(String, CanonValue)]) -> CanonValue {
    let mut converted: Vec<(CanonValue, CanonValue)> = pairs
        .iter()
        .map(|(k, v)| (CanonValue::keyword(k.clone()), v.clone()))
        .collect();
    // CanonValue::map sorts pairs by encoded key bytes internally, but sorting
    // here keeps the preimage deterministic regardless.
    converted.sort_by_key(|a| a.0.encode());
    CanonValue::map(converted)
}

/// Round-lifecycle root: domain hash of the round-lifecycle projection under
/// ROUND_LIFECYCLE_V1.
pub fn round_lifecycle_root(lifecycle: &CanonValue) -> String {
    domain_hash("ROUND_LIFECYCLE_V1", lifecycle)
}

/// Canonical ROUND_LIFECYCLE_V1 preimage for a `RoundLifecycle`, mirroring the
/// Clojure `round-state/round-lifecycle` map field-for-field.
pub fn lifecycle_canon_value(lc: &RoundLifecycle) -> CanonValue {
    let mut pairs: Vec<(CanonValue, CanonValue)> = vec![
        (
            CanonValue::keyword("assurance"),
            CanonValue::str(lc.assurance.clone()),
        ),
        (
            CanonValue::keyword("cancellation-blocking-reasons"),
            CanonValue::array(
                lc.cancellation_blocking_reasons
                    .iter()
                    .map(|s| CanonValue::str(s.clone()))
                    .collect(),
            ),
        ),
        (
            CanonValue::keyword("cancellation-possible"),
            CanonValue::bool(lc.cancellation_possible),
        ),
        (
            CanonValue::keyword("cancellation-window"),
            CanonValue::str(lc.cancellation_window.clone()),
        ),
        (
            CanonValue::keyword("cancellation-window-schema"),
            CanonValue::str(lc.cancellation_window_schema.clone()),
        ),
        (
            CanonValue::keyword("derived-state"),
            match &lc.derived_state {
                Some(s) => CanonValue::str(s.clone()),
                None => CanonValue::null(),
            },
        ),
        (
            CanonValue::keyword("evidence-status"),
            CanonValue::str(lc.evidence_status.clone()),
        ),
        (
            CanonValue::keyword("lifecycle-assertion-status"),
            CanonValue::str(lc.lifecycle_assertion_status.clone()),
        ),
        (
            CanonValue::keyword("lifecycle-profile-id"),
            CanonValue::str(lc.lifecycle_profile_id.clone()),
        ),
        (
            CanonValue::keyword("lifecycle-profile-version"),
            CanonValue::int(num_bigint::BigInt::from(lc.lifecycle_profile_version)),
        ),
        (
            CanonValue::keyword("round-state"),
            match &lc.round_state {
                Some(s) => CanonValue::str(s.clone()),
                None => CanonValue::null(),
            },
        ),
    ];
    pairs.sort_by_key(|(k, _)| k.encode());
    CanonValue::map(pairs)
}

/// Realized-results root: domain hash of the realized-results disposition
/// vector under REALIZED_RESULTS_V1.
pub fn realized_results_root(realized: &RealizedAllocation) -> String {
    domain_hash("REALIZED_RESULTS_V1", &realized.canon_results())
}

/// The canonical statement preimage (the value committed by the statement root).
pub fn statement_preimage(
    allocation_context_root: &str,
    request_set_root: &str,
    allocation_policy_root: &str,
    realized_results_root: &str,
    fail_action_policy_root: &str,
    round_lifecycle_root: &str,
) -> CanonValue {
    CanonValue::map(vec![
        (
            CanonValue::keyword("schema-version"),
            CanonValue::str(SCHEMA_VERSION),
        ),
        (
            CanonValue::keyword("allocation-context-root"),
            CanonValue::str(allocation_context_root),
        ),
        (
            CanonValue::keyword("request-set-root"),
            CanonValue::str(request_set_root),
        ),
        (
            CanonValue::keyword("allocation-policy-root"),
            CanonValue::str(allocation_policy_root),
        ),
        (
            CanonValue::keyword("realized-results-root"),
            CanonValue::str(realized_results_root),
        ),
        (
            CanonValue::keyword("fail-action-policy-root"),
            CanonValue::str(fail_action_policy_root),
        ),
        (
            CanonValue::keyword("round-lifecycle-root"),
            CanonValue::str(round_lifecycle_root),
        ),
    ])
}

/// The canonical statement root: domain hash of the statement preimage under
/// REALIZED_ALLOCATION_STATEMENT_V1.
pub fn statement_root(
    allocation_context_root: &str,
    request_set_root: &str,
    allocation_policy_root: &str,
    realized_results_root: &str,
    fail_action_policy_root: &str,
    round_lifecycle_root: &str,
) -> String {
    domain_hash(
        "REALIZED_ALLOCATION_STATEMENT_V1",
        &statement_preimage(
            allocation_context_root,
            request_set_root,
            allocation_policy_root,
            realized_results_root,
            fail_action_policy_root,
            round_lifecycle_root,
        ),
    )
}

/// A fully projected realized-allocation statement.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct RealizedAllocationStatement {
    pub schema_version: String,
    pub allocation_context_root: String,
    pub request_set_root: String,
    pub allocation_policy_root: String,
    pub realized_results_root: String,
    pub fail_action_policy_root: String,
    pub round_lifecycle_root: String,
    pub statement_root: String,
    pub all_active: bool,
}

/// Build a realized-allocation-statement.v1 from raw inputs, recomputing all
/// six roots.
///
/// Args:
///   - `ctx`                — the parsed allocation context (raw round inputs)
///   - `requested`          — claim-id -> requested amount (sorted)
///   - `policy`             — effective fill policy key/value pairs (minus
///     :fail-action-policy)
///   - `fail_action_policy` — declared fail-action policy, or None for the
///     conservative default
///   - `lifecycle`          — canonical round-lifecycle projection
///   - `realized`           — the realized allocation (semantics layer output)
pub fn build_statement(
    ctx: &Context,
    requested: &[(String, i64)],
    policy: &[(String, CanonValue)],
    fail_action_policy: Option<&[(String, CanonValue)]>,
    lifecycle: &CanonValue,
    realized: &RealizedAllocation,
) -> RealizedAllocationStatement {
    let acr = allocation_context_root(ctx);
    let rsr = request_set_root(requested);
    let apr = allocation_policy_root(policy);
    let rrr = realized_results_root(realized);
    let fap = fail_action_policy_root(fail_action_policy);
    let rl = round_lifecycle_root(lifecycle);
    let sr = statement_root(&acr, &rsr, &apr, &rrr, &fap, &rl);
    let all_active = realized
        .participants
        .iter()
        .all(|p| p.filled == p.requested && p.deferred == 0 && p.haircut == 0);
    RealizedAllocationStatement {
        schema_version: SCHEMA_VERSION.to_string(),
        allocation_context_root: acr,
        request_set_root: rsr,
        allocation_policy_root: apr,
        realized_results_root: rrr,
        fail_action_policy_root: fap,
        round_lifecycle_root: rl,
        statement_root: sr,
        all_active,
    }
}

/// Helpers for constructing canonical policy maps.
pub mod policy {
    use crate::canonical::CanonValue;

    /// Build a fill-policy map from (key, keyword-value) pairs, e.g.
    /// [("mode", "pro-rata"), ("rounding-policy", "largest-remainder")].
    pub fn keyword_policy(pairs: &[(&str, &str)]) -> Vec<(String, CanonValue)> {
        pairs
            .iter()
            .map(|(k, v)| ((*k).to_string(), CanonValue::keyword(*v)))
            .collect()
    }
}

/// Minimal public-input helpers for exposing the statement (what SP1 should
/// reveal): the canonical statement root plus whichever public inputs are
/// intentionally exposed. The scenario-evidence binding stays simulator-side.
pub fn reveal(statement: &RealizedAllocationStatement, public_inputs: &[&str]) -> CanonValue {
    CanonValue::map(
        std::iter::once((
            CanonValue::keyword("statement-root"),
            CanonValue::str(statement.statement_root.clone()),
        ))
        .chain(
            public_inputs
                .iter()
                .map(|k| (CanonValue::keyword(*k), CanonValue::str(*k))),
        )
        .collect(),
    )
}
